File: solutions/java/src/main/java/com/epi/LongestValidParentheses.java
package com.epi;

import java.util.Deque;
import java.util.LinkedList;

public class LongestValidParentheses {
  // @include
  public static int longestValidParentheses(String s) {
    int maxLength = 0, end = -1;
    Deque<Integer> leftParenthesesIndices = new LinkedList<>();
    for (int i = 0; i < s.length(); ++i) {
      if (s.charAt(i) == '(') {
        leftParenthesesIndices.addFirst(i);
      } else if (leftParenthesesIndices.isEmpty()) {
        end = i;
      } else {
        leftParenthesesIndices.removeFirst();
        int start = leftParenthesesIndices.isEmpty()
                        ? end
                        : leftParenthesesIndices.peekFirst();
        maxLength = Math.max(maxLength, i - start);
      }
    }
    return maxLength;
  }
  // @exclude

  private static void smallTest() {
    assert(longestValidParentheses(")(((())()(()(") == 6);
    assert(longestValidParentheses("((())()(()(") == 6);
    assert(longestValidParentheses(")(") == 0);
    assert(longestValidParentheses("()") == 2);
    assert(longestValidParentheses("") == 0);
    assert(longestValidParentheses("()()())") == 6);
  }

  public static void main(String[] args) {
    smallTest();
    if (args.length == 1) {
      String s = args[0];
      System.out.println("s = " + s);
      System.out.println(longestValidParentheses(s));
    }
  }
}

File: solutions/cpp/longest-valid-parentheses.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <stack>
#include <string>
#include <vector>

using std::cout;
using std::endl;
using std::max;
using std::stack;
using std::string;
using std::vector;

// @include
int LongestValidParentheses(const string& s) {
  int max_length = 0, end = -1;
  stack<int> left_parentheses_indices;
  for (int i = 0; i < s.size(); ++i) {
    if (s[i] == '(') {
      left_parentheses_indices.emplace(i);
    } else if (left_parentheses_indices.empty()) {
      end = i;
    } else {
      left_parentheses_indices.pop();
      int start = left_parentheses_indices.empty()
                      ? end
                      : left_parentheses_indices.top();
      max_length = max(max_length, i - start);
    }
  }
  return max_length;
}
// @exclude

void SmallTest() {
  assert(LongestValidParentheses(")(((())()(()(") == 6);
  assert(LongestValidParentheses("((())()(()(") == 6);
  assert(LongestValidParentheses(")(") == 0);
  assert(LongestValidParentheses("()") == 2);
  assert(LongestValidParentheses("") == 0);
  assert(LongestValidParentheses("()()())") == 6);
}

int main(int argc, char** argv) {
  SmallTest();
  string s;
  if (argc == 2) {
    s = argv[1];
  }
  cout << "s = " << s << endl;
  cout << LongestValidParentheses(s) << endl;
  return 0;
}
